File: quarkdsl-web/lib/vm/ast.ts
/**
 * QuarkDSL Abstract Syntax Tree (AST) Types
 * Represents the hierarchical structure of parsed programs
 */

export enum Domain {
  Classical = "Classical",
  Gpu = "Gpu",
  Quantum = "Quantum",
}

export enum TypeKind {
  Int = "Int",
  Float = "Float",
  Bool = "Bool",
  Qubit = "Qubit",
  Void = "Void",
  QState = "QState",
  Array = "Array",
  Tensor = "Tensor",
}

export interface Type {
  kind: TypeKind;
  elementType?: Type;
  size?: number;
}

export interface Param {
  name: string;
  ty: Type;
}

export interface Function {
  name: string;
  params: Param[];
  returnType: Type;
  body: Statement[];
  domain: Domain;
}

export interface Program {
  functions: Function[];
}

export enum BinaryOp {
  Add = "Add",
  Sub = "Sub",
  Mul = "Mul",
  Div = "Div",
  Mod = "Mod",
  Eq = "Eq",
  Ne = "Ne",
  Lt = "Lt",
  Le = "Le",
  Gt = "Gt",
  Ge = "Ge",
  And = "And",
  Or = "Or",
}

export enum UnaryOp {
  Neg = "Neg",
  Not = "Not",
}

export type Expression =
  | { kind: "IntLiteral"; value: number }
  | { kind: "FloatLiteral"; value: number }
  | { kind: "BoolLiteral"; value: boolean }
  | { kind: "Variable"; name: string }
  | { kind: "Binary"; op: BinaryOp; left: Expression; right: Expression }
  | { kind: "Unary"; op: UnaryOp; operand: Expression }
  | { kind: "Call"; function: string; args: Expression[] }
  | { kind: "Index"; array: Expression; index: Expression }
  | { kind: "ArrayLiteral"; elements: Expression[] }
  | { kind: "Map"; function: string; array: Expression };

export type Statement =
  | { kind: "Let"; name: string; ty?: Type; value: Expression }
  | { kind: "Assign"; target: string; index?: Expression; value: Expression }
  | { kind: "Return"; value: Expression }
  | { kind: "Expression"; expr: Expression }
  | { kind: "For"; var: string; start: Expression; end: Expression; body: Statement[] }
  | { kind: "If"; condition: Expression; thenBody: Statement[]; elseBody?: Statement[] };

File: src/frontend/ast.rs
/// Abstract Syntax Tree definitions for QuarkDSL

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq)]
pub struct Program {
    pub functions: Vec<Function>,
}

/// Execution domain for functions
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum Domain {
    Classical,  // CPU execution (default)
    Gpu,        // GPU execution (@gpu)
    Quantum,    // Quantum execution (@quantum)
}

#[derive(Debug, Clone, PartialEq)]
pub struct Function {
    pub name: String,
    pub params: Vec<Param>,
    pub return_type: Type,
    pub body: Vec<Statement>,
    pub domain: Domain,  // NEW: execution domain
}

#[derive(Debug, Clone, PartialEq)]
pub struct Param {
    pub name: String,
    pub ty: Type,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Type {
    Int,
    Float,
    Bool,
    Array(Box<Type>, Option<usize>), // element type, optional size
    Qubit,
    Void,
    Tensor(Box<Type>),  // NEW: GPU tensor type, e.g., tensor<float>
    QState,             // NEW: Quantum state type
}

#[derive(Debug, Clone, PartialEq)]
pub enum Statement {
    Let {
        name: String,
        ty: Option<Type>,
        value: Expression,
    },
    Assign {
        target: String,
        index: Option<Box<Expression>>,
        value: Expression,
    },
    Return(Expression),
    Expression(Expression),
    For {
        var: String,
        start: Expression,
        end: Expression,
        body: Vec<Statement>,
    },
    If {
        condition: Expression,
        then_body: Vec<Statement>,
        else_body: Option<Vec<Statement>>,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub enum Expression {
    IntLiteral(i64),
    FloatLiteral(f64),
    BoolLiteral(bool),
    Variable(String),
    ArrayLiteral(Vec<Expression>),
    Index {
        array: Box<Expression>,
        index: Box<Expression>,
    },
    Binary {
        op: BinaryOp,
        left: Box<Expression>,
        right: Box<Expression>,
    },
    Unary {
        op: UnaryOp,
        operand: Box<Expression>,
    },
    Call {
        function: String,
        args: Vec<Expression>,
    },
    Map {
        function: String,
        array: Box<Expression>,
    },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BinaryOp {
    Add,
    Sub,
    Mul,
    Div,
    Mod,
    Eq,
    Ne,
    Lt,
    Le,
    Gt,
    Ge,
    And,
    Or,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum UnaryOp {
    Neg,
    Not,
}

impl std::fmt::Display for Type {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Type::Int => write!(f, "int"),
            Type::Float => write!(f, "float"),
            Type::Bool => write!(f, "bool"),
            Type::Array(elem, Some(size)) => write!(f, "[{}; {}]", elem, size),
            Type::Array(elem, None) => write!(f, "[{}]", elem),
            Type::Qubit => write!(f, "qubit"),
            Type::Void => write!(f, "void"),
            Type::Tensor(elem) => write!(f, "tensor<{}>", elem),
            Type::QState => write!(f, "qstate"),
        }
    }
}

